src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/DependencyAnalysis/EETypeNode.cs

```diff
@@ -457,11 +461,22 @@ public sealed override IEnumerable<CombinedDependencyListEntry> GetConditionalSt
                         if (!isStaticInterfaceMethod && !needsDependenciesForInstanceInterfaceMethodImpls)
                             continue;
 
+                        MethodDesc interfaceMethodDefinition = interfaceMethod;
+                        if (interfaceType != interfaceDefinitionType)
+                            interfaceMethodDefinition = factory.TypeSystemContext.GetMethodForInstantiatedType(interfaceMethodDefinition.GetTypicalMethodDefinition(), (InstantiatedType)interfaceDefinitionType);
+
                         MethodDesc implMethod = isStaticInterfaceMethod ?
-                            defType.ResolveInterfaceMethodToStaticVirtualMethodOnType(interfaceMethod) :
-                            defType.ResolveInterfaceMethodToVirtualMethodOnType(interfaceMethod);
+                            defTypeDefinition.ResolveInterfaceMethodToStaticVirtualMethodOnType(interfaceMethodDefinition) :
+                            defTypeDefinition.ResolveInterfaceMethodToVirtualMethodOnType(interfaceMethodDefinition);
                         if (implMethod != null)
                         {
+                            TypeDesc implType = defType;
+                            while (!implType.HasSameTypeDefinition(implMethod.OwningType))
+                                implType = implType.BaseType;
+
+                            if (!implType.IsTypeDefinition)
+                                implMethod = factory.TypeSystemContext.GetMethodForInstantiatedType(implMethod.GetTypicalMethodDefinition(), (InstantiatedType)implType);
```

Comment:
And very much not sure what we're doing here.

Reply:
Given:

```csharp
interface IFoo<T>
{
    void M(T t);
}

class Foo<T, U> : IFoo<U>
{
    public void M(U t) { }
}
```

And computing the conditional dependencies for `Foo<object, object>`:

The above logic just resolved that `IFoo<!1>.M` is implemented by `Foo<T, U>.M(U)`. We now need to find this method on `Foo<object, object>`. This grabs the typical method definition of `Foo<T, U>.M(U)` (the same thing we have right now) and finds it on `Foo<object, object>`, returning `Foo<object, object>(object)`

This code is here mostly to cover the more interesting case of:

```csharp
class Base<U>
{
    public void M(U t) { }
}

class Foo<T, U> : Base<U>, IFoo<U>
{
}
```

Where the result of the resolution would be `Base<!1>.M` and we want `Base<object>.M`.